v2transport/transport.go

```diff
@@ -0,0 +1,625 @@
+	// unique sessionID.
+	sessionInfo := []byte("session_id")
+	sessionReader := hkdf.Expand(sha256.New, prk, sessionInfo)
+	_, err := sessionReader.Read(p.sessionID)
+	if err != nil {
+		return err
+	}
+
+	// Use the Expand operation to generate packet and packet-length encryption
+	// ciphers.
+	initiatorLInfo := []byte("initiator_L")
+	initiatorLReader := hkdf.Expand(sha256.New, prk, initiatorLInfo)
+	_, err = initiatorLReader.Read(p.initiatorL)
+	if err != nil {
+		return err
+	}
+
+	initiatorPInfo := []byte("initiator_P")
+	initiatorPReader := hkdf.Expand(sha256.New, prk, initiatorPInfo)
+	_, err = initiatorPReader.Read(p.initiatorP)
+	if err != nil {
+		return err
+	}
+
+	responderLInfo := []byte("responder_L")
+	responderLReader := hkdf.Expand(sha256.New, prk, responderLInfo)
+	_, err = responderLReader.Read(p.responderL)
+	if err != nil {
+		return err
+	}
+
+	responderPInfo := []byte("responder_P")
+	responderPReader := hkdf.Expand(sha256.New, prk, responderPInfo)
+	_, err = responderPReader.Read(p.responderP)
+	if err != nil {
+		return err
+	}
+
+	// Create the garbage terminators that each side will use.
+	garbageInfo := []byte("garbage_terminators")
+	garbageReader := hkdf.Expand(sha256.New, prk, garbageInfo)
+	garbageTerminators := make([]byte, 32)
+	_, err = garbageReader.Read(garbageTerminators)
+	if err != nil {
+		return err
+	}
+
+	initiatorGarbageTerminator := garbageTerminators[:garbageSize]
+	responderGarbageTerminator := garbageTerminators[garbageSize:]
+
+	if initiating {
+		p.sendL, err = NewFSChaCha20(p.initiatorL)
+		if err != nil {
+			return err
+		}
+
+		p.sendP, err = NewFSChaCha20Poly1305(p.initiatorP)
+		if err != nil {
+			return err
+		}
+
+		p.sendGarbageTerm = initiatorGarbageTerminator
+
+		p.recvL, err = NewFSChaCha20(p.responderL)
+		if err != nil {
+			return err
+		}
+
+		p.recvP, err = NewFSChaCha20Poly1305(p.responderP)
+		if err != nil {
+			return err
+		}
+
+		p.recvGarbageTerm = responderGarbageTerminator
+	} else {
+		p.sendL, err = NewFSChaCha20(p.responderL)
+		if err != nil {
+			return err
+		}
+
+		p.sendP, err = NewFSChaCha20Poly1305(p.responderP)
+		if err != nil {
+			return err
+		}
+
+		p.sendGarbageTerm = responderGarbageTerminator
+
+		p.recvL, err = NewFSChaCha20(p.initiatorL)
+		if err != nil {
+			return err
+		}
+
+		p.recvP, err = NewFSChaCha20Poly1305(p.initiatorP)
+		if err != nil {
+			return err
+		}
+
+		p.recvGarbageTerm = initiatorGarbageTerminator
+	}
+
+	// TODO:
+	//     To achieve forward secrecy we must wipe the key material used to initialize the ciphers:
+	//     memory_cleanse(ecdh_secret, prk, initiator_L, initiator_P, responder_L, responder_K)
+	// - golang analogue?
+
+	return nil
+}
+
+// initiateV2Handshake generates our private key and sends our public key as
+// well as garbage data to our peer.
+func (p *Peer) InitiateV2Handshake(garbageLen int) error {
+	var err error
+	p.privkeyOurs, p.ellswiftOurs, err = btcec.EllswiftCreate()
```

Comment:
Thought this was in a sub-package? If not, we should move it to one.

Reply:
Will move